src/Symfony/Component/Validator/Constraints/UrlValidator.php

```diff
@@ -62,7 +62,16 @@ public function validate($value, Constraint $constraint)
             return;
         }
 
-        $pattern = $constraint->relativeProtocol ? str_replace('(%s):', '(?:(%s):)?', static::PATTERN) : static::PATTERN;
+        $pattern = static::PATTERN;
+        if ($constraint->allowExtraChars) {
+            $pattern = str_replace(
+                '(?:\? (?:[\pL\pN\-._\~!$&\'()*+,;=:@/?]|%%[0-9A-Fa-f]{2})* )?',
+                '(?:\? (?:[\pL\pN\-._\~!$&\'\[\]()*+,;=:@/?]|%%[0-9A-Fa-f]{2})* )?',
+                $pattern
+            );
+        }
+
+        $pattern = $constraint->relativeProtocol ? str_replace('(%s):', '(?:(%s):)?', $pattern) : $pattern;
         $pattern = sprintf($pattern, implode('|', $constraint->protocols));
```

Comment:
my idea was to put an extra `%s` in the pattern, and provide it here: `preg_quote($constraint->allowedExtraChars, '~')`

Then a user could do e.g. `allowedExtraChars = '[]`, or any char. But maybe it's an edge case that only applies to brackets, in that case i would call the option `allowBrackets`.

Reply:
you're right, allowBrackets seems more understandable